src/esp/gfx/DebugLineRender.cpp

```diff
@@ -0,0 +1,314 @@
+}
+
+void DebugLineRender::drawBox(const Magnum::Vector3& min,
+                              const Magnum::Vector3& max,
+                              const Magnum::Color4& color) {
+  // 4 lines along x axis
+  drawTransformedLine(Mn::Vector3(min.x(), min.y(), min.z()),
+                      Mn::Vector3(max.x(), min.y(), min.z()), color);
+  drawTransformedLine(Mn::Vector3(min.x(), min.y(), max.z()),
+                      Mn::Vector3(max.x(), min.y(), max.z()), color);
+  drawTransformedLine(Mn::Vector3(min.x(), max.y(), min.z()),
+                      Mn::Vector3(max.x(), max.y(), min.z()), color);
+  drawTransformedLine(Mn::Vector3(min.x(), max.y(), max.z()),
+                      Mn::Vector3(max.x(), max.y(), max.z()), color);
+
+  // 4 lines along y axis
+  drawTransformedLine(Mn::Vector3(min.x(), min.y(), min.z()),
+                      Mn::Vector3(min.x(), max.y(), min.z()), color);
+  drawTransformedLine(Mn::Vector3(max.x(), min.y(), min.z()),
+                      Mn::Vector3(max.x(), max.y(), min.z()), color);
+  drawTransformedLine(Mn::Vector3(min.x(), min.y(), max.z()),
+                      Mn::Vector3(min.x(), max.y(), max.z()), color);
+  drawTransformedLine(Mn::Vector3(max.x(), min.y(), max.z()),
+                      Mn::Vector3(max.x(), max.y(), max.z()), color);
+
+  // 4 lines along z axis
+  drawTransformedLine(Mn::Vector3(min.x(), min.y(), min.z()),
+                      Mn::Vector3(min.x(), min.y(), max.z()), color);
+  drawTransformedLine(Mn::Vector3(max.x(), min.y(), min.z()),
+                      Mn::Vector3(max.x(), min.y(), max.z()), color);
+  drawTransformedLine(Mn::Vector3(min.x(), max.y(), min.z()),
+                      Mn::Vector3(min.x(), max.y(), max.z()), color);
+  drawTransformedLine(Mn::Vector3(max.x(), max.y(), min.z()),
+                      Mn::Vector3(max.x(), max.y(), max.z()), color);
+}
+
+void DebugLineRender::drawCircle(const Magnum::Vector3& pos,
+                                 float radius,
+                                 const Magnum::Color4& color,
+                                 int numSegments,
+                                 const Magnum::Vector3& normal) {
+  // https://stackoverflow.com/questions/11132681/what-is-a-formula-to-get-a-vector-perpendicular-to-another-vector
+  auto randomPerpVec = normal.z() < normal.x()
+                           ? Mn::Vector3(normal.y(), -normal.x(), 0)
+                           : Mn::Vector3(0, -normal.z(), normal.y());
+
+  pushTransform(Mn::Matrix4::lookAt(pos, pos + normal, randomPerpVec) *
+                Mn::Matrix4::scaling(Mn::Vector3(radius, radius, 0.f)));
+
+  Mn::Vector3 prevPt;
+  for (int seg = 0; seg <= numSegments; seg++) {
+    Mn::Deg angle = Mn::Deg(360.f * float(seg) / numSegments);
+    Mn::Vector3 pt(Mn::Math::cos(angle), Mn::Math::sin(angle), 0.f);
+    if (seg > 0) {
+      drawTransformedLine(prevPt, pt, color);
+    }
+    prevPt = pt;
+  }
+
+  popTransform();
+}
```

Comment:
No need to reinvent anything :) 

I was thinking you could have a generic `drawPrimitive()` utility:

```cpp
void DebugLineRender::drawPrimitive(const Mn::Trade::MeshData& mesh, const Mn::Matrix4& transformation, const Mn::Color4& color) {
  /* Turn the mesh into nonindexed lines first */
  if(mesh.primitive() != Mn::MeshPrimitive::Lines) {
    CORRADE_INTERNAL_ASSERT(mesh.primitive() == Mn::MeshPrimitive::LineLoop ||
                            mesh.primitive() == Mn::MeshPrimitive::LineStrip);
    return drawPrimitive(Mn::MeshTools::generateIndices(mesh), transformation, color);
  }
  if(mesh.isIndexed())
    return drawPrimitive(Mn::MeshTools::duplicate(mesh), transformation, color);

  /* Steal all the positions */
  for(const Mn::Vector3& position: mesh.attribute<Mn::Vector3>(Mn::Trade::MeshAttribute::Position))
    arrayAppend(_verts, Cr::InPlaceInit, transformation.transformPoint(position), color);
}
```

and then just reuse whatever existing primitive code (didn't you also have a debug renderer with colored axes? there's [`Primitives::axis3D()`](https://doc.magnum.graphics/magnum/namespaceMagnum_1_1Primitives.html#a4c8248d7f389b072c9d1a6a25ee363b0) for that too):

```cpp
void drawBox(const Mn::Vector3& min, const Mn::Vector3& max, const Mn::Color4& color) {
  drawPrimitive(Mn::Primitives::cubeWireframe(),
    _cachedInputTransform*
    Mn::Matrix4::translation((max + min)/2.0f)* /* i hope this is correct? */
    Mn::Matrix4::scaling((max - min)/2.0f),
    color);
}

void drawCircle(const Mn::Vector3& pos, float radius, int segments, const Mn::Color4& color, const Mn::Vector3& normal) {
  ...
  drawPrimitive(Mn::Primitives::circle3DWireframe(segments),
    _cachedInputTransform*
    Mn::Matrix4::lookAt(pos, pos + normal, randomPerpVec)*
    Mn::Matrix4::scaling(Mn::Vector3(radius, radius, 0.f)));
}
```

Add `#include`s as needed. The `Primitives` library should be sufficiently lightweight to have negligible perf impact when used this way.

Reply:
I feel like your suggestion here is more complicated than what I have. From my perspective, a bit of code to create a circle or a box is simple and readable, and users can easily adapt it later to their own shapes (tetrahedrons, helices, etc.).

Reply:
The code has a lot of extra API calls with long names to massage the mesh data, yes, but the essence of it is a single loop that copies line endpoints from a large gallery of existing primitives, where adding another one (such as a cone for visualizing spotlights) is extremely trivial :)

I commented also because my past experience is that any primitive generation code that wasn't thoroughly tested tended to rot over time, getting increasingly broken during minor refactorings. Since I suppose you won't want to spend time writing tests ensuring the mesh is generated correctly, this would delegate that responsibility to the engine.

Reply:
For now, rather than drag out the discussion, let's agree to disagree. :)

I'll add a comment in the file pointing to this discussion, in case folks in the future want to go this route.